### PR TITLE
ChatOps Collector PR

### PR DESCRIPTION
**Pull request by amitmawkin**
Supports Enterprise Hipchat implementation

### CODE REVIEW

File: Chat-Ops-Collector/pom.xml [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    <properties>
+        <java.version>1.7</java.version>
+    </properties>
+
+    <dependencies>
+        <!-- Project Deps -->
+        <dependency>
+            <groupId>com.capitalone.dashboard</groupId>
+            <artifactId>core</artifactId>
+            <version>1.5.0-SNAPSHOT</version>
```
Comment (jdamick) [inline Chat-Ops-Collector/pom.xml]: <version>${project.version}</version>

File: Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/collector/ChatOpsCollectorTask.java [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * CollectorTask that fetches Commit information from GitHub
+ */
+@Component
+public class ChatOpsCollectorTask extends CollectorTask<Collector> {
+    private static final Log LOG = LogFactory.getLog(ChatOpsCollectorTask.class);
```
Comment (jdamick) [inline Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/collector/ChatOpsCollectorTask.java]: switch to slf4j to match the others..

File: Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/collector/ChatOpsCollectorTask.java [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+            token3 = StringUtils.leftPad(elapsed, 30 - text.length() );
+        } else {
+            String countStr = count.toString();
+            token2 = StringUtils.leftPad(countStr, 20 - text.length() );
+            token3 = StringUtils.leftPad(elapsed, 10 );
+        }
+        LOG.info(text + token2 + token3);
+    }
+
+    protected void logBanner(String instanceUrl) {
```
Comment (jdamick) [inline Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/collector/ChatOpsCollectorTask.java]: these are inherited now, you can just use those..

File: Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/collector/ChatOpsCollectorTask.java [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+        return chatOpsRepository.findEnabledChatOpsRepos(collector.getId());
+    }
+
+ 
+
+    protected void log(String marker, long start) {
+        log(marker, start, null);
+    }
+
+    protected void log(String text, long start, Integer count) {
```
Comment (jdamick) [inline Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/collector/ChatOpsCollectorTask.java]: these are inherited now, you can just use those..

File: Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/collector/ChatOpsCollectorTask.java [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+        return repo.getLastUpdateTime();
+    }
+
+    private List<ChatOpsRepo> enabledRepos(Collector collector) {
+        return chatOpsRepository.findEnabledChatOpsRepos(collector.getId());
+    }
+
+ 
+
+    protected void log(String marker, long start) {
```
Comment (jdamick) [inline Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/collector/ChatOpsCollectorTask.java]: these are inherited now, you can just use those..

File: Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/collector/DefaultChatOpsClient.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+import org.springframework.stereotype.Component;
+import org.springframework.web.client.RestOperations;
+
+/**
+ * ChatOpsClient
+ */
+
+@Component
+public class DefaultChatOpsClient implements ChatOpsClient {
+	private static final Log LOG = LogFactory.getLog(DefaultChatOpsClient.class);
```
Comment (jdamick) [inline Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/collector/DefaultChatOpsClient.java]: change to slf4j

File: Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/collector/RestOperationsSupplier.java [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+import org.springframework.stereotype.Component;
+import org.springframework.web.client.RestOperations;
+import org.springframework.web.client.RestTemplate;
+
+@Component
+public class RestOperationsSupplier implements Supplier<RestOperations> {
+    @Override
+    public RestOperations get() {
+        HttpComponentsClientHttpRequestFactory requestFactory = new HttpComponentsClientHttpRequestFactory();
+        requestFactory.setConnectTimeout(20000);
```
Comment (jdamick) [inline Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/collector/RestOperationsSupplier.java]: these timeouts should be configurable..

File: Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/model/ChatOpsRepo.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
+package com.capitalone.dashboard.model;
+
+import org.joda.time.DateTime;
+
+/**
+ * CollectorItem extension to store the chatops serverUrl, roomName and authToken.
+ */
+public class ChatOpsRepo extends CollectorItem {
+    private static final String CHATOPS_SERVER_URL = "chatOpsServerUrl";
```
Comment (jdamick) [inline Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/model/ChatOpsRepo.java]: this might be better as an enum..

File: Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/model/ChatOpsRepo.java [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+        if (this == o) {
+        	return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+        	return false;
+        }
+
+        ChatOpsRepo chatOpsRepo = (ChatOpsRepo) o;
+
+        return getChatopsServerUrl().equals(chatOpsRepo.getChatopsServerUrl()) & getChatopsServerUrl().equals(chatOpsRepo.getChatopsServerUrl()) & getChatopsAuthToken().equals(chatOpsRepo.getChatopsAuthToken());
```
Comment (jdamick) [inline Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/model/ChatOpsRepo.java]: use Objects.equals(getChatopsServerUrl(), chatOpsRepo.getChatopsServerUrl()), it handles case of nulls..

File: Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/model/ChatOpsRepo.java [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    @Override
+    public int hashCode() {
+        return getChatopsServerUrl().hashCode();
```
Comment (jdamick) [inline Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/model/ChatOpsRepo.java]: Objects.hashCode(getChatopsServerUrl(), getChatopsAuthToken(), getChatopsRoomName());

File: Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/model/GitHubRepo.java [hunk 11]
```diff
@@ -0,0 +1,57 @@
+package com.capitalone.dashboard.model;
+
+import org.joda.time.DateTime;
+
+/**
+ * CollectorItem extension to store the github repo url and branch.
+ */
+public class GitHubRepo extends CollectorItem {
```
Comment (jdamick) [inline Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/model/GitHubRepo.java]: why is githubrepo copied into here?

File: Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/repository/GitHubRepoRepository.java [hunk 12]
```diff
@@ -0,0 +1,16 @@
+package com.capitalone.dashboard.repository;
+
+import com.capitalone.dashboard.model.GitHubRepo;
+import org.bson.types.ObjectId;
+import org.springframework.data.mongodb.repository.Query;
+
+import java.util.List;
+
+public interface GitHubRepoRepository extends BaseCollectorItemRepository<GitHubRepo> {
```
Comment (jdamick) [inline Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/repository/GitHubRepoRepository.java]: why is this here?

File: UI/src/app/dashboard/core/data-factories/chatops-data.js [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+        };
+
+        function details(serviceUrl) {
+            console.log("ServiceURl:"+serviceUrl);
+            return $http.get(serviceUrl).then(function (response) {
+
+                    return response.data;
+                },function(response){
+                console.log("Error occured:"+JSON.stringify(response));
+                return response.data;
```
Comment (jdamick) [inline UI/src/app/dashboard/core/data-factories/chatops-data.js]: ?

File: UI/src/components/widgets/chatops/config.js [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+
+        if (!widgetConfig.options.chatOpsOption) {
+            ctrl.chatOpsOption = "";
+        }
+        else {
+            var myindex;
+
+            for (var v = 0; v < ctrl.chatOpsOptions.length; v++) {
+                if (ctrl.chatOpsOptions[v].name == widgetConfig.options.chatOpsOption.name) {
+                    myindex = v;
```
Comment (jdamick) [inline UI/src/components/widgets/chatops/config.js]: why continue to loop after u find it?

File: UI/src/components/widgets/chatops/view.js [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+        console.log("Complete URl is :" + completeUrl);
+
+
+        ctrl.load = function () {
+            console.log("In load");
+            var deferred = $q.defer();
+
+            chatOpsData.details(completeUrl).then(function (data) {
+
+                console.log("DATA GOT BACK:" + JSON.stringify(data));
```
Comment (jdamick) [inline UI/src/components/widgets/chatops/view.js]: remove

File: UI/src/components/widgets/chatops/view.js [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+            }
+            if (min < 10) {
+                min = "0" + min;
+            }
+            var messdates = hour + ":" + min + ":" + sec;
+            return messdates;
+        };
+
+
+        ctrl.getImageUrl = function (hipchatuserid) {
```
Comment (jdamick) [inline UI/src/components/widgets/chatops/view.js]: i'm not sure if we should do this..

**Comment by jdamick** [comment]
rather than "collecting" all of the chats - i think we should just embed the room in the UI: 
For example: 
<https://bitbucket.org/atlassianlabs/jquery-hipchat-plugin>

I just dont think it makes sense to pull in all of that ..

File: Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/repository/GitHubRepoRepository.java (hunk 12 above)
Comment (amitmawkin) [inline Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/repository/GitHubRepoRepository.java]: I thought I removed it, not sure why my fork did not get updated with my commit, as I used the Github collector as a template

File: Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/collector/ChatOpsCollectorTask.java (hunk 2 above)
Comment (amitmawkin) [inline Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/collector/ChatOpsCollectorTask.java]: other have common logging
import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;

File: Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/collector/ChatOpsCollectorTask.java (hunk 3 above)
Comment (amitmawkin) [inline Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/collector/ChatOpsCollectorTask.java]: removed the redundant method

File: Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/collector/ChatOpsCollectorTask.java (hunk 2 above)
Comment (amitmawkin) [inline Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/collector/ChatOpsCollectorTask.java]: i removed it all together do not need it,

File: Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/collector/ChatOpsCollectorTask.java (hunk 4 above)
Comment (amitmawkin) [inline Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/collector/ChatOpsCollectorTask.java]: got it, used all the inherited one's

File: Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/collector/ChatOpsCollectorTask.java (hunk 5 above)
Comment (amitmawkin) [inline Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/collector/ChatOpsCollectorTask.java]: Used Inherited

File: Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/collector/DefaultChatOpsClient.java (hunk 6 above)
Comment (amitmawkin) [inline Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/collector/DefaultChatOpsClient.java]: removed the logger

File: Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/collector/RestOperationsSupplier.java (hunk 7 above)
Comment (amitmawkin) [inline Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/collector/RestOperationsSupplier.java]: will have to refactor all the collectors, it's on a to do list with the admin page story

File: Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/model/ChatOpsRepo.java (hunk 8 above)
Comment (amitmawkin) [inline Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/model/ChatOpsRepo.java]: This is a dummy field, once api is updated to register a widget of this type it will not be needed anymore

File: Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/model/GitHubRepo.java (hunk 11 above)
Comment (amitmawkin) [inline Chat-Ops-Collector/src/main/java/com/capitalone/dashboard/model/GitHubRepo.java]: removed...

File: UI/src/components/widgets/chatops/config.js (hunk 14 above)
Comment (amitmawkin) [inline UI/src/components/widgets/chatops/config.js]: added a break;

File: UI/src/components/widgets/chatops/view.js (hunk 15 above)
Comment (amitmawkin) [inline UI/src/components/widgets/chatops/view.js]: removed

File: UI/src/components/widgets/chatops/view.js (hunk 16 above)
Comment (amitmawkin) [inline UI/src/components/widgets/chatops/view.js]: its not needed any more, I was testing something else...

**Comment by jdamick** [comment]
i really dont think we should support this - and i think we should revert.
